src/wallet/coin_selection.rs

```diff
@@ -165,6 +186,7 @@ pub trait CoinSelectionAlgorithm<D: Database>: std::fmt::Debug {
         fee_rate: FeeRate,
         amount_needed: u64,
         fee_amount: u64,
+        _drain_output: &TxOut,
```

Comment:
Technically here you don't need the full `TxOut`, the `Script` should be enough and I think it would be less confusing because it makes more sense that your coin selection depends on the change script, not some kind of output.

I would change this to a `drain_script` or something, then you'll have to replace the size estimation for the txout and it should be `serialize(script).len() + 8` (please double check this).

It's very important that you `serialize` the script and avoid using `Script::len()`: the former should also prefix the script with its length in VarInt format, which I believe is how it gets encoded in a `TxOut`. The latter only tells you the length of the script.

Reply:
Totally, this is great!
I double checked that `serialize(drain_output).len() == (serialize(drain_script).len() + 8)` and you're right.
Both are equal to: `size(len(script_pubkey)) + len(script_pubkey) + size(output_value)` where
`size(len(script_pubkey))` is the size of the `VarInt` field used to describe the length of the script.